### PR TITLE
Only get processes info if pid is running [fix #1303]

### LINKED ISSUE #1303: `processes` table pid constraint leads to non existent rows
Querying the processes table with a query that generates a pid constraint gives interesting results for pid's of processes that are not running.  In the case of a process that isn't running we would expect no row to be generated.  Instead we see:

```
osquery> SELECT * FROM processes WHERE pid=123;
I0706 09:27:46.494313 1961374464 processes.cpp:175] An error occurred retrieving the env for pid: 123
           pid = 123
          name =
          path =
       cmdline =
           cwd =
          root =
           uid = -1
           gid = -1
          euid = -1
          egid = -1
       on_disk = -1
    wired_size = -1
 resident_size = -1
phys_footprint = -1
     user_time = -1
   system_time = -1
    start_time = -1
        parent = -1
```

This is the same for OS X and linux despite the separate implementations.  Thoughts?  This doesn't seem like a big deal.

### PR DESCRIPTION
**Pull request by marpaia**

### CODE REVIEW

**Comment by mofarrell** [comment]
This is also an issue in linux builds.

**Comment by marpaia** [comment]
@mofarrell good catch, i'll add it for linux as well.

**Comment by theopolis** [comment]
(1) What is the motivation for the additional header/implementation that is shared among both Linux/Darwin platforms? The implementation seems very platform specific such that it'll be more useful to just include alongside either's `processes.cpp`, which already contains lots of useful process-related utility methods. 
(2) Are we sure the processes table is the only one affected?

**Comment by mofarrell** [comment]
Just wanted to revoice my two concerns from https://github.com/marpaia/osquery/commit/493a66bd709bf45544177fd6b34b3a421672e56a#comments

**Comment by marpaia** [comment]
@theopolis agreed with 1, I changed them back to occur in the individual table implementations. With regards to 2, I updated the other process related table by moving out the pid acquisition into a utility and using it from the table generators for the other tables.

@mofarrell gotcha, fair point, I added a constraint before we iterate the pid list that ensures that the size of the constraints is less than the total active pids. This is easy on OS X, because we have the call to `proc_listpids` which returns the count of items without returning the actual pids. On linux, we're just parsing `/proc`, so there's a bit of a caveat. I thought about adding another function which returns the count of pids, but since the implementation would be exactly the same complexity, I think it makes sense to count the number of pids in `/proc` by normally iterating `/proc`, then compare the `size()` of the `set` with the number of elements in the constraint list. if there are in fact less constraints, we clear the `set`, re-using the same data structure and re-fill it with the smaller, verified subset of pids.

**Comment by marpaia** [comment]
@osqueryer test this please

File: osquery/tables/system/linux/processes.cpp [hunk 1]
```diff
@@ -56,6 +56,26 @@ inline std::string readProcLink(const std::string& attr, const std::string& pid)
   return result;
 }
 
+Status getPidsForContext(QueryContext& context, std::set<std::string>& pids) {
+  auto s = osquery::procProcesses(pids);
```
Comment (theopolis) [inline osquery/tables/system/linux/processes.cpp]: I'm interested in the latency incurred by visiting all leafs in the `/proc` directory.

File: osquery/tables/system/linux/processes.cpp [hunk 2]
```diff
@@ -56,6 +56,26 @@ inline std::string readProcLink(const std::string& attr, const std::string& pid)
...
+    return s;
+  }
+
+  if (context.constraints.count("pid") > 0 &&
+      context.constraints["pid"].exists(EQUALS) &&
+      context.constraints.at("pid").getAll(EQUALS).size() < pids.size()) {
+    pids.clear();
+    for (const auto& pid :
+         context.constraints["pid"].getAll<std::string>(EQUALS)) {
+      if (isDirectory("/proc/" + pid).ok()) {
```
Comment (theopolis) [inline osquery/tables/system/linux/processes.cpp]: Don't you already know if the pid was in the `pids` set? What if you construct a new set if the constraint exists and move matching pids over? That will be considerably less work than `stat`ing the directory.

File: osquery/tables/system/darwin/processes.cpp [hunk 3]
```diff
@@ -33,11 +33,40 @@ namespace tables {
...
+    VLOG(1) << "An error occurred retrieving the process list";
+    return pidlist;
+  }
+
   if (context.constraints.count("pid") > 0 &&
-      context.constraints.at("pid").exists(EQUALS)) {
-    pidlist = context.constraints.at("pid").getAll<int>(EQUALS);
+      context.constraints.at("pid").exists(EQUALS) &&
+      context.constraints.at("pid").getAll<int>(EQUALS).size() <
+          (bufsize / sizeof(pid_t))) {
```
Comment (theopolis) [inline osquery/tables/system/darwin/processes.cpp]: Maybe a `auto& constraint = context.constraints.at("pid")` for a more readable `if` statement? Might not make a difference.

**Pull request closed without merging**